### PR TITLE
Handle edge case

### PR DESCRIPTION
**Pull request by jcrugzz**
## Summary

The goal here is to support an edge case as follows.

We have a rootPkg `x@10.0.0` with a subpackage `y@10.0.0`.

We publish `x@11.0.0` and `y@11.0.0` as new major versions.

We then go and publish `x@10.1.0` which will auto publish `y@10.0.1-0`.

But now if we go to publish `x@11.1.0`, before this change, we would not have properly auto published `11.0.1-0`, we would have not triggered sub packages because we would naively think that this was the first time major version 11 appeared. 

This is the use case fixed by the changes made.

## Changelog

* Fix edge case for current -> previous -> current package publishes with proper auto triggering of dependent builds

## Test Plan

Added test and made the rest of the tests create the proper data that was needed for the test we added.

### CODE REVIEW

File: lib/feedsme/index.js [hunk 1]
```diff
@@ -316,12 +362,15 @@ Feedsme.prototype._triggerOneDep = async function _triggerOneDep({ publish, prom
...
  */
-Feedsme.prototype._triggerStrategy = function _triggerStrategy({ releaseLine, rootPkg, pkg, env }) {
+Feedsme.prototype._triggerStrategy = function _triggerStrategy({ registryRootPkg, releaseLine, rootPkg, pkg, env }) {
   if (env !== 'dev' && releaseLine) return { strategy: 'release', trigger: true };
   if (!releaseLine) return { strategy: 'legacy', trigger: true };
 
   const { name, version: rootVersion } = rootPkg;
   const { dependencies } = pkg;
+  const rootPkgVersions = Object.keys(registryRootPkg.versions || {})
+    .sort((y, x) => semver.compare(x, y));
```
Comment (msluther) [inline lib/feedsme/index.js]: I assume this is intentionally inverting `x` & `y`?

File: lib/feedsme/index.js [hunk 2]
```diff
@@ -566,6 +628,22 @@ Feedsme.prototype._respec = function respec(key) {
...
+/**
+ * Fetch a given package name from the configured registry
+ *
+ * @param {String} name - Package name
+ * @returns {Promise<Object>} Wrapped pkg object
+ */
+Feedsme.prototype._fetchFromRegistry = function (name) {
+  return new Promise((resolve, reject) => {
+    const uri = `${this.registryUrl}/${encodeURIComponent(name)}`;
+    this.registry.get(uri, this.params,  (err, pkg) => {
```
Comment (msluther) [inline lib/feedsme/index.js]: should you just [`promisify`](https://nodejs.org/dist/latest-v10.x/docs/api/util.html#util_util_promisify_original) that instead of building your own promise?

File: lib/feedsme/index.js [hunk 3]
```diff
@@ -40,6 +41,27 @@ function Feedsme({ agents, models, config, log }) {
...
+      randomize: true
+    }
+  });
+
+  const npm = config.get('npm');
+  const { auth, ...parsed } = url.parse(npm);
+  this.registryUrl = url.format(parsed).replace(/\/$/, '');
+  const [username, password] = (auth || '').split(':');
+
+  this.params = {
```
Comment (msluther) [inline lib/feedsme/index.js]: maybe `this.registryParams` ?

File: lib/feedsme/index.js [hunk 4]
```diff
@@ -40,6 +41,27 @@ function Feedsme({ agents, models, config, log }) {
...
+  this.registry = new NpmRegistry({
+    retry: {
+      count: 3,
+      factor: 2,
+      randomize: true
+    }
+  });
+
+  const npm = config.get('npm');
+  const { auth, ...parsed } = url.parse(npm);
```
Comment (msluther) [inline lib/feedsme/index.js:54]: Should this support token auth as well or just username/password?

File: lib/feedsme/index.js [hunk 5]
```diff
@@ -316,12 +362,15 @@ Feedsme.prototype._triggerOneDep = async function _triggerOneDep({ publish, prom
...
  * @api private
  */
-Feedsme.prototype._triggerStrategy = function _triggerStrategy({ releaseLine, rootPkg, pkg, env }) {
+Feedsme.prototype._triggerStrategy = function _triggerStrategy({ registryRootPkg, releaseLine, rootPkg, pkg, env }) {
   if (env !== 'dev' && releaseLine) return { strategy: 'release', trigger: true };
   if (!releaseLine) return { strategy: 'legacy', trigger: true };
 
   const { name, version: rootVersion } = rootPkg;
   const { dependencies } = pkg;
+  const rootPkgVersions = Object.keys(registryRootPkg.versions || {})
```
Comment (msluther) [inline lib/feedsme/index.js]: you've got this `|| {}` here, but then never actually check existence of `latestRootPkgVersion` before passing it into `semver.major` below. is this just extra sanity checking that should never happen 🤞 or should you be checking that some versions exist before using it (or defaulting to `0.0.0`)?  I'm assuming we don't want `semver.major` to throw

File: lib/feedsme/index.js [hunk 6]
```diff
@@ -344,14 +393,23 @@ Feedsme.prototype._triggerStrategy = function _triggerStrategy({ releaseLine, ro
   // increment so it doesnt mess with future publishes
   //
   const previousMajor = semver.major(rootVersion) < semver.major(rootDepVersion);
-
-  const trigger = previousMajor ? true : inclusive;
+  const previousPublishedMajor = semver.major(rootVersion) === semver.major(latestRootPkgVersion);
```
Comment (msluther) [inline lib/feedsme/index.js:407]: some comments here would be good. because this reads really weird. Explain how this test means that it's a previous published version.  From outside this PR it seems like this would be just saying, "is the current root version the latest published version to npm" when wouldn't this comparison here be true?

File: lib/feedsme/index.js (hunk 4 above)
Comment (indexzero) [reply lib/feedsme/index.js:54]: Just username & password are supported today through most of the Warehouse tooling. Sounds like a good cross-cutting concern we should calibrate on though.

File: lib/feedsme/index.js [hunk 7]
```diff
@@ -316,12 +362,15 @@ Feedsme.prototype._triggerOneDep = async function _triggerOneDep({ publish, prom
...
-Feedsme.prototype._triggerStrategy = function _triggerStrategy({ releaseLine, rootPkg, pkg, env }) {
+Feedsme.prototype._triggerStrategy = function _triggerStrategy({ registryRootPkg, releaseLine, rootPkg, pkg, env }) {
   if (env !== 'dev' && releaseLine) return { strategy: 'release', trigger: true };
   if (!releaseLine) return { strategy: 'legacy', trigger: true };
 
   const { name, version: rootVersion } = rootPkg;
   const { dependencies } = pkg;
+  const rootPkgVersions = Object.keys(registryRootPkg.versions || {})
+    .sort((y, x) => semver.compare(x, y));
+  const latestRootPkgVersion = rootPkgVersions[0];
```
Comment (indexzero) [inline lib/feedsme/index.js]: This would be more obvious using `.find` imho:

``` js
const latestRootPkgVersion = Object.keys(registryRootPkg.versions || {})
  // Sort in descending order 
  .sort((x, y) => semver.compare(y, x))
  // Return latest version 
  .find(v => v);
```

File: lib/feedsme/index.js (hunk 1 above)
Comment (indexzero) [reply lib/feedsme/index.js]: Yes, since it is sorting in descending order – but a comment would be helpful. Also I'd suggest inverting the invert (see my other comment below) 

To prove the descending order point ... e.g.

``` js
const versions = [ '0.0.0',
  '0.1.0',
  '0.2.0',
  '0.3.0',
  '0.3.1',
  '0.3.2',
  '0.3.3',
  '0.4.0',
  '0.5.0',
  '0.6.0',
  '0.7.0',
  '0.8.0',
  '0.8.1',
  '0.8.2',
  '1.0.0',
  '1.1.0',
  '1.1.1',
  '1.1.2',
  '1.1.3',
  '1.1.4',
  '1.1.5',
  '1.2.0',
  '1.2.1',
  '1.2.2',
  '1.2.3',
  '1.3.0',
  '1.4.0',
  '2.0.0' ];

console.log(versions.sort((y, x) => semver.compare(x, y));
```

which outputs:

```
[ '2.0.0',
  '1.4.0',
  '1.3.0',
  '1.2.3',
  '1.2.2',
  '1.2.1',
  '1.2.0',
  '1.1.5',
  '1.1.4',
  '1.1.3',
  '1.1.2',
  '1.1.1',
  '1.1.0',
  '1.0.0',
  '0.8.2',
  '0.8.1',
  '0.8.0',
  '0.7.0',
  '0.6.0',
  '0.5.0',
  '0.4.0',
  '0.3.3',
  '0.3.2',
  '0.3.1',
  '0.3.0',
  '0.2.0',
  '0.1.0',
  '0.0.0' ]
```

File: lib/feedsme/index.js (hunk 6 above)
Comment (indexzero) [reply lib/feedsme/index.js:407]: The names for these variables need an overhaul – agreed on that point. In in this case the "root" name is confusing since it is somewhat interchangeable with "parent" but it's unclear when "pkg" is a dependent package or not.

That said, I think we should simply open an issue outlining this confusion and then do a pass after this ships. Objections?

File: test/feedsme.test.js [hunk 8]
```diff
@@ -462,167 +483,282 @@ describe('feedsme', function () {
...
-        await Promise.all([
-          Package.create(latest),
-          Version.create({
-            versionId: `${latest.name}@${latest.version}`,
-            name: latest.name,
-            version: latest.version,
-            value: JSON.stringify(body)
-          })
-        ]);
+      describe('sequence with huh and what packages', function () {
```
Comment (indexzero) [inline test/feedsme.test.js:486]: We **really** need better naming conventions and documentation around these scenarios. I believe the best path forward is to start making true [sequence diagrams](https://mermaidjs.github.io/sequenceDiagram.html) for these scenarios. UML isn't always the answer, but sequence diagrams were designed for communicating exactly these kinds of time-based edge cases. 

If we had a sequence diagram for each `describe` we would have a high level of confidence in our test coverage. Reducing the unknown unknowns.

**Review by indexzero — APPROVED** [review]
Would like to see the comments addressed, but 👍 once they are addressed. New test coverage is 💯

File: lib/feedsme/index.js (hunk 6 above)
Comment (msluther) [reply lib/feedsme/index.js:407]: I think we should add a comment about what we're doing here now, and open an issue to come back and revisit the function and clean it up along with addressing the new comment about no longer needing to walk the release lines since we have the full published version set now.

File: lib/feedsme/index.js (hunk 1 above)
Comment (jcrugzz) [reply lib/feedsme/index.js]: @indexzero @msluther i guess we could use `rcompare` to make it more clear.

File: lib/feedsme/index.js (hunk 5 above)
Comment (jcrugzz) [reply lib/feedsme/index.js]: @msluther yea this actually should never happen. I can remove this or be even more defensive. Im not sure if registry.get actually errors on a 404. if it does this would 100% never happen, if not then being more defensive cant hurt.

File: lib/feedsme/index.js (hunk 6 above)
Comment (jcrugzz) [reply lib/feedsme/index.js:407]: Commented but yea this code is confusing :(

File: lib/feedsme/index.js (hunk 7 above)
Comment (jcrugzz) [reply lib/feedsme/index.js]: huh, i guess thats not so bad. Changed to this and used `rcompare` to make it more clear.

File: test/feedsme.test.js (hunk 8 above)
Comment (jcrugzz) [reply test/feedsme.test.js:486]: Agreed.

**Review by indexzero — APPROVED** [review]
Follow-up 👍 – comments appear to be addressed.

**Review by msluther — APPROVED** [review]
